### PR TITLE
Zephyr Build info file

### PR DESCRIPTION
**Pull request by tejlmand**
This PR introduces a common build info file for Zephyr.

Some tools peeks into Zephyr's CMakeCache.txt file and relies on what is considered internal build variables.
This is bad for several reasons, especially as it suddenly puts unknown expectation on the Zephyr build system.
And it makes it harder to cleanup Zephyr's build system when we are doing clean ups in the build code.

Besides the lack of visibility, such as which variables are used externally / downstream, then it's also hard for developers that are integrating to Zephyr to know which settings they can expect to be constant.

Some settings are documented as input to Zephyr build system, like EXTRA_CONF_FILE, EXTRA_DTS_OVERLAY_FILE, etc. but are used together with internal logic on conf file selection, making it hard to extract knowledge on configuration file.

#78727 also present a use-case where internal build knowledge must be exported to Python

To support various use-cases, then this PR proposes a common interface which allows to store build values in a common yaml file.
The yaml file is proposed to be named and placed in `<build-dir>/build_info.yml`.

A yaml schema is provided in https://github.com/zephyrproject-rtos/zephyr/blob/82dd714dc8dbcf0b1fa138cd521117ec49e3a2be/scripts/schemas/build-schema.yml.
This schema acts as the contract regarding information from the build system.
For example, the schema contains:
```
  cmake:
    type: map
    mapping:
      ...
      kconfig:
        type: map
        mapping:
          files:
            type: seq
            sequence:
              - type: str
```
thus promising to place kconfig files in use under this key.
The build system are the free to have this information in an internal variable and rename this variable as it find fit.

To facilitate populating the build info file, then a new CMake function is created: `build_info(<key> VALUE <value>)`
Example:
```
  build_info(kconfig user-files VALUE ${CONF_FILE_AS_LIST})
```
this allows the code to call `build_info()` close to the final usage of the internal CMake variable.
Furthermore, having a `build_info()` function makes it easy to find internal settings that are exported, and any CMake variables expanded in the `<value>` field can easily be renamed if build system internals are changed.

If `build_info()` is called with an invalid key, like `build_info(does not exists VALUE ok)`, then an error is thrown.
This both helps to ensure that only keys described in the schema are allowed as well as catching typos.

It also ensures that in the event of a key is renamed / removed then usage of said key in `build_info()` will raise an error.

To both minimize the need for Zephyr modules to require n-entries in the schema as well as supporting vendor specific information, then a free form `vendor-specific` section is provided. This section is considered under the responsibilities of the vendors using it.
https://github.com/zephyrproject-rtos/zephyr/blob/82dd714dc8dbcf0b1fa138cd521117ec49e3a2be/scripts/schemas/build-schema.yml#L92-L99

### CODE REVIEW

File: scripts/west_commands/build.py
Comment (pdgendt) [inline scripts/west_commands/build.py:244]: Add the `west` version too.

Comment (tejlmand) [reply scripts/west_commands/build.py:244]: good point

Comment (pdgendt) [reply scripts/west_commands/build.py:244]: I guess this is true for almost all entries?

Version for cmake/python/...

Comment (tejlmand) [reply scripts/west_commands/build.py:244]: yep, providing the version of each tool we describe in the build info file is good.

**Comment by pdgendt** [comment]
Any reason not to use a `.yml/.yaml` extension? Free syntax highlighting.

**Comment by pdgendt** [comment]
Some comments when running

```shell
west build -p always -b native_sim samples/hello_world
# output snippets
  toolchain:
    name: host
    path: 
  zephyr:
    version: 3.7.99
```

- The toolchain info isn't really useful
- Use/add `git_describe` from `git.cmake` output for zephyr/application

**Comment by tejlmand** [comment]
> Any reason not to use a .yml/.yaml extension? Free syntax highlighting.

Only reason was to be somewhat aligned with the default `zephyr.hex`, `zephyr.elf`, `zephyr.bin`, ... files which gave `zephyr.build` as to indicate the purpose of the file at the expense of file extension. (The alternative `zephyr.yml` doesn't say much about purpose).

Naming the file `build.yaml` has the risk of being confused with a "true" build file, such as `build.ninja`.
But i'm open to good naming suggestions.

**Comment by pdgendt** [comment]
> Naming the file `build.yaml` has the risk of being confused with a "true" build file, such as `build.ninja`. But i'm open to good naming suggestions.

`build_info.yaml` would be my suggestion, showing a direct link from the function to the output.

**Comment by tejlmand** [comment]
> Some comments when running

Thanks, appreciated :+1: 

> - Duplicate entries in the kconfig file lists

Correct, and the reason for this is to provide a list of all files which are used for the kconfig, but also a dedicated list of those files which are considered user / application controlled and thus directly under the developers control.
As example, a normal developer should be editing `prj.conf`, `board/<board-target>.conf`, files given by `EXTRA_CONF_FILE`, etc, but not files provided by the board (those should be updated by the board maintainer), files created by sysbuild, etc.

One may argue that external tool can concatenate those lists but the tricky part is that internally in CMake all files are being appended to a common new list later at which point we would need to split the lists / rework how they are constructed.
Thus I decided to keep a list of all files, and the list of user facing files.

> - The toolchain info isn't really useful

True for the `host` toolchain, but when you start to be using different toolchains, such as Zehyr SDK, LLVM for Arm, Arm compiler 6 (armclang), then this info actually has value.
Also for an IDE which may want to show which toolchain was selected.

> - Use/add git_describe from git.cmake output for zephyr/application

Doing so will have other implications, so should not be part of this PR. If we want this, then a dedicated issue / PR for this should be opened.
Some short reasons why this should initially be left out is because the build info is generated at CMake configure time and if any information currently populated in the CMake section changes then CMake will be automatically re-run.
However a regular incremental build, that is where you change a source file, for example `main.c`, commit the changes and then do an incremental build will not trigger a CMake re-run, which means that the git describe will not match what is actually build (because the commit just changed the SHA of the repo, but CMake did not re-run to update the build info).
See also: https://github.com/zephyrproject-rtos/zephyr/pull/42527 and https://github.com/zephyrproject-rtos/zephyr/issues/40167 as to why we should avoid re-running CMake on every git commit / incremental build.

Secondly, for the git SHAs we have the build meta file for Zephyr which already generates an output file at build time containing the SHA of Zephyr as well as each Zephyr module, so that feature covers much more than just a git describe of Zephyr itself.
See: 
https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_BUILD_OUTPUT_META
https://docs.zephyrproject.org/latest/develop/west/zephyr-cmds.html#software-bill-of-materials-west-spdx

So no reason to duplicate information in the build info which we already in a safer way produces at build time.

**Review by nordicjm — COMMENTED** [review]
Some thoughts:

* Comments refer to `build.info` file but it's actually `zephyr.build`.
* In a sysbuild project (smp_svr) it seems to have the same file in multiple places:
```
   kconfig:
    files:
     - /tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/build/_sysbuild/empty.conf
     - /tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/sysbuild.conf
    user-files:
     - /tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/sysbuild.conf
```
and
```
  kconfig:
    files:
     - /tmp/aa/zephyr/boards/nordic/nrf5340dk/nrf5340dk_nrf5340_cpuapp_defconfig
     - /tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/prj.conf
     - /tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/build/smp_svr/zephyr/.config.sysbuild
    user-files:
     - /tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/prj.conf
```
and
```
  devicetree:
    bindings-dirs:
     - /tmp/aa/zephyr/dts/bindings
    files:
     - /tmp/aa/zephyr/boards/nordic/nrf5340dk/nrf5340dk_nrf5340_cpuapp.dts
     - /tmp/aa/bootloader/mcuboot/boot/zephyr/app.overlay
    include-dirs:
     - /tmp/aa/bootloader/mcuboot/boot/zephyr/include
...
    user-files:
     - /tmp/aa/bootloader/mcuboot/boot/zephyr/app.overlay
```
* Another oddity:
```
  kconfig:
    files:
     - /tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/_AA/_sysbuild/empty.conf
     - /tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/_AA/_sysbuild/empty.conf
     - /tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/sysbuild.conf
    user-files:
     - /tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/sysbuild.conf
```
* Would it make sense to also have the original zephyr commit as well as the version, so that it can be tracked too?
* This might include the files that were used but does not mean a build can be reproduced e.g. if a user includes their own command line overrides or if they used FILE_SUFFIX

**Comment by tejlmand** [comment]
@nordicjm  thanks for the comments.

 > Comments refer to build.info file but it's actually zephyr.build.

you mean commit messages ?
Will correct that to `build_info.yaml` if no other names are proposed, see also ref: https://github.com/zephyrproject-rtos/zephyr/pull/79118#issuecomment-2379599804

> In a sysbuild project (smp_svr) it seems to have the same file in multiple places:

See explanation here: https://github.com/zephyrproject-rtos/zephyr/pull/79118#issuecomment-2379908948

> Another oddity:

yes, this is a sysbuild / kconfig reuse issue. The current `kconfig.cmake` implementation requires a given number of files (even if they are empty), so sysbuild simply provides references to empty files to make Kconfig happy.

This has nothing to do with build info but becomes more visible now.
Cleanup of kconfig.cmake in this regard is outside scope of build info PR and should be addressed separately.

> Would it make sense to also have the original zephyr commit as well as the version, so that it can be tracked too?

See reply here https://github.com/zephyrproject-rtos/zephyr/pull/79118#issuecomment-2379908948 why git commit or similar should be left out initially.

> This might include the files that were used but does not mean a build can be reproduced e.g. if a user includes their own command line overrides or if they used FILE_SUFFIX

It's reproducible in the sense that you can see exactly what was used by the build system, but not in the sense that you have an explicit command that will produce exactly the list of files, as example if extra Zephyr modules are available.

The purpose of this PR is to provide a solution which shows what was used, for example to be used by an IDE to present configuration files, or a downstream project which needs to use settings for own build invocation, see: #78727 for an example.

Trying to provide a solution which provides a single command for reproducible builds is a different task, because a user might manually modify CMakeCache.txt after the first CMake invocation, but you have no way to know what changed and if a given setting can just be appended to original CMake invocation.
As example, running `cmake -DFOO=y` + `modify CMakeCache` and add `BAR=y` may not behave the same as `cmake -DFOO=y -DBAR=y` because a given cache setting `BAZ` could default in first run based on the value of `BAR`, but if `BAR` is later changed then `BAZ` may be sticky (FORCE vs no FORCE when writing cache value).
This is especially true in Kconfig wrt. stuck symbols: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#stuck-symbols-in-menuconfig-and-guiconfig.

Please pay attention to the description of the PR:
> > This PR introduces a common build info file for Zephyr.
> >
> > Some tools peeks into Zephyr's CMakeCache.txt file and relies on what is considered internal build variables.
> > This is bad for several reasons, especially as it suddenly puts unknown expectation on the Zephyr build system.
> > And it makes it harder to cleanup Zephyr's build system when we are doing clean ups in the build code.

what is useful with this PR in relation to reproducible builds is the fact that you now have a file with the internal info from a given build which you may compare against a known expectation, and if the file you generate doesn't match expectation then you can compare and see what the difference are, which is much easier than comparing two elf files or two build.ninja files.
It might even be beneficial during a `git bisect` session as you have this file after CMake run, before the build stage.

And combined with build meta feature we already have: 
https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_BUILD_OUTPUT_META
https://docs.zephyrproject.org/latest/develop/west/zephyr-cmds.html#software-bill-of-materials-west-spdx

then we are one more step closer to really be able to provide real support for reproducible builds.

**Comment by tejlmand** [comment]
> `build_info.yaml` would be my suggestion, showing a direct link from the function to the output.

Done.

File: scripts/west_commands/build.py
Comment (pdgendt) [inline scripts/west_commands/build.py:26]:
```suggestion
BUILD_INFO_LOG = 'build_info.yaml'
```

Comment (tejlmand) [reply scripts/west_commands/build.py]: thanks for catching. Fixed.

**Review by pdgendt — DISMISSED** [review]

File: scripts/west_commands/build.py
Comment (tejlmand) [reply scripts/west_commands/build.py]: added `west` version in this PR.

Support for a list of tools and their versions will follow in a new PR.
Reason is I want to provide those as:
```
  tools:
    - name: python
      version: 3.10.4
    - name: CMake
      version: 3.24.5
```

however that requires that we guarantee this limitation is fixed:
https://github.com/zephyrproject-rtos/zephyr/blob/df7f1ff5b5b5cd18c68ccb676ca210b82b9fe6d4/cmake/modules/yaml.cmake#L32-L38

**Comment by pillo79** [comment]
Thanks for this, super useful, already contains almost all the info I was looking forward to add in #78727. :+1: 
The only bit I'm still missing from both the LLEXT EDK and this PR is the _target triple_, as defined by the toolchain (AFAICS that's `CROSS_COMPILE_TARGET` for gcc and  `CMAKE_C_COMPILER_TARGET` for clang). 

I would really like to see that being exported, as the current toolchain information is not sufficient to identify the actual compiler being used. This bit of information is pretty useful to anyone trying to use the EDK with the Zephyr SDK.

**Comment by tejlmand** [comment]
> The only bit I'm still missing from both the LLEXT EDK and this PR is the target triple, as defined by the toolchain (AFAICS that's CROSS_COMPILE_TARGET for gcc and CMAKE_C_COMPILER_TARGET for clang).
>
> I would really like to see that being exported, as the current toolchain information is not sufficient to identify the actual compiler being used. This bit of information is pretty useful to anyone trying to use the EDK with the Zephyr SDK.

Thanks for having a look and give feedback.
If it is only the triple we are missing then I can add them in this PR, else I considered this PR the baseline and we can create extra commits  / PRs with all the additional info to add to build info, whatever you prefer :slightly_smiling_face: 

There is already plan for an upcoming `tools` section as discussed here: https://github.com/zephyrproject-rtos/zephyr/pull/79118#discussion_r1782369557

**Comment by pillo79** [comment]
With this PR in, the way the EDK exported information from the build can be simplified. If you think it's OK to also add the final list of compilation flags to `build_info.yaml`, I can amend the EDK PR after this one is merged. :+1:

**Comment by tejlmand** [comment]
> With this PR in, the way the EDK exported information from the build can be simplified.

Great to hear :slightly_smiling_face: 
> If you think it's OK to also add the final list of compilation flags to build_info.yaml,

I think this is a perfect valid wish. Just one thing I realized is that exporting compilation flags to the build_info requires a small enhancement to current solution. Reason is that compilation flags can be set **after** `find_package(Zephyr)`, meaning after `build_info` has been generated. This requires that we adjust build info to be created / updated at build time instead of configure time (current solution).

Will look at this enhancement asap, but let get this PR in as a baseline and extend from there.

> I can amend the EDK PR after this one is merged. 👍

:+1: sound like a good plan.

**Review by pillo79 — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
@tejlmand can you rebase please?

**Comment by nashif** [comment]
this is showing some issues that we need to fix...

```
    bindings-dirs:
     - /home/nashif/zephyrproject/zephyr/dts/bindings
    files:
     - /home/nashif/zephyrproject/zephyr/boards/qemu/x86/qemu_x86_atom_nopae.dts
    include-dirs:
     - /home/nashif/zephyrproject/modules/hal/atmel/include
     - /home/nashif/zephyrproject/modules/hal/gigadevice/include
     - /home/nashif/zephyrproject/modules/hal/microchip/dts
     - /home/nashif/zephyrproject/modules/hal/nuvoton/dts
     - /home/nashif/zephyrproject/modules/hal/nxp/dts
     - /home/nashif/zephyrproject/modules/hal/stm32/dts
     - /home/nashif/zephyrproject/zephyr/include
     - /home/nashif/zephyrproject/zephyr/include/zephyr
     - /home/nashif/zephyrproject/zephyr/dts/common
     - /home/nashif/zephyrproject/zephyr/dts/x86
     - /home/nashif/zephyrproject/zephyr/dts/xtensa
     - /home/nashif/zephyrproject/zephyr/dts/sparc
     - /home/nashif/zephyrproject/zephyr/dts/riscv
     - /home/nashif/zephyrproject/zephyr/dts/posix
     - /home/nashif/zephyrproject/zephyr/dts/nios2
     - /home/nashif/zephyrproject/zephyr/dts/arm64
     - /home/nashif/zephyrproject/zephyr/dts/arm
     - /home/nashif/zephyrproject/zephyr/dts/arc
     - /home/nashif/zephyrproject/zephyr/dts
```

why are we getting includes for unrelated modules and hals? looks like those are just being added for each build?

**Comment by tejlmand** [comment]
> this is showing some issues that we need to fix...

agree, but fixing up this is outside the scope of build info.
Build info just reveals what is already there.

**Comment by nashif** [comment]
> agree, but fixing this is outside the scope of build info.
> Build info just reveals what is already there.

right, it seems the dts part is coming from

```
dts_root: .
```
in some of the hals.

**Comment by nashif** [comment]
```
cmake:
  application:
    configuration-dir: /home/nashif/zephyrproject/zephyr/samples/hello_world
    source-dir: /home/nashif/zephyrproject/zephyr/samples/hello_world
  board:
    name: qemu_x86
    qualifiers: /atom
    revision:
  devicetree:
    bindings-dirs:
     - /home/nashif/zephyrproject/zephyr/dts/bindings
     
```

why do the qualifiers start with '/'?  IMO this is wrong and misleading and conflict with what we have in the docs, the "/" is how you connect the various elements, not part of the qualifiers themselves.

**Comment by tejlmand** [comment]
> why do the qualifiers start with '/'? IMO this is wrong and misleading and conflict with what we have in the docs,

Thanks for catching, fixed :+1:

**Review by pdgendt — APPROVED** [review]

**Review by pillo79 — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]

File: cmake/modules/FindHostTools.cmake
Comment (WilliamGFish) [inline cmake/modules/FindHostTools.cmake:119]: Causing old compilation errors as ???_TOOLCHAIN_PATH is invalid. 

> -- Found toolchain: gnuarmemb (C:\gnu_arm_embedded_10)
> CMake Error at C:/Users/BillyBob/zephyrGoodX/zephyrNew/cmake/modules/yaml.cmake:313 (string):
>   string sub-command JSON failed parsing json string: * Line 1, Column 1
> 
>     Bad escape sequence in string
> 
>   See Line 1, Column 6 for detail.


I changed it to:
```
build_info(toolchain path VALUE "${TOOLCHAIN_ROOT}")
# build_info(toolchain path VALUE "${${zephyr_toolchain_variant_upper}_TOOLCHAIN_PATH}")
```

Comment (fabiobaltieri) [reply cmake/modules/FindHostTools.cmake:119]: Hey @WilliamGFish, could you open a pull request with the fix? This is the [contribution guideline documentation](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-guidelines).

Comment (tejlmand) [reply cmake/modules/FindHostTools.cmake:119]: a couple of comments regarding this, first related to the issue itself.

Please note that a CMake path officially is with forward slashes `/`, and that windows path separator in principle should be escaped.
This info seems to be a bit hidden in various places of the docs, like here:
https://cmake.org/cmake/help/latest/command/cmake_path.html#modification
> If `<input>` is a native path, it is converted into a cmake-style path with forward-slashes (/).
> ...
> ...
> For example:
>  `set(native_path "c:\\a\\b/..\\c")`

or in CMake questions:
https://discourse.cmake.org/t/window-env-variable-and-cmake-path/5410
> Paths should normally be written with forward slashes in CMake code, even for Windows. Alternatively, if you want a backslash, you need to escape it, so your string would have to be either `"D:/test/fmt/release"` or `"D:\\test\\fmt\\release"`.

So the use of a single back-slash in general with CMake is **not** guaranteed to work.
That said, it seems like the `find_program()` supports the use of single `\`. The `<toolchain>_TOOLCHAIN_PATH` is generally used together with `find_program()` and therefore working.

Regarding the `build_info()` / yaml feature, which uses CMake's json support under the hood, then path should of course be properly sanitized in case they contain single `\`.

Note that `TOOLCHAIN_ROOT` != `<toolchain>_TOOLCHAIN_PATH`.
One is the toolchain itself, the other points to the location of the toolchain CMake integration code.
Those two variables points to very different things.

Comment (tejlmand) [reply cmake/modules/FindHostTools.cmake:119]: Fixed here: https://github.com/zephyrproject-rtos/zephyr/pull/79647

Comment (WilliamGFish) [reply cmake/modules/FindHostTools.cmake:119]: You're right they are two different things just needed stop compile failures. I have addressed this by replacing the \ with / in the CMake script like this:
```
string(REGEX REPLACE "\\\\" "/" toolchain_path "${${zephyr_toolchain_variant_upper}_TOOLCHAIN_PATH}")
build_info(toolchain path VALUE "${toolchain_path}")
```
As on windows machines they will continue to use backslashes in config files the will need to be replaced.

**Pull request merged**
